src/lib/codemirror-editor.ts

```diff
@@ -96,13 +128,31 @@ export class CodeMirrorEditorElement extends LitElement {
    * If true, display a left-hand-side gutter with line numbers. Default false
    * (hidden).
    */
-  @property({type: Boolean, attribute: 'line-numbers'})
+  @property({type: Boolean, attribute: 'line-numbers', reflect: true})
   lineNumbers = false;
 
+  /**
+   * If true, this editor is not editable.
+   */
+  @property({type: Boolean, reflect: true})
+  readonly = false;
+
+  /**
+   * The CodeMirror theme to load.
+   */
+  @property()
+  theme = 'default';
+
   private _resizeObserver?: ResizeObserver;
 
   update(changedProperties: PropertyValues) {
-    if (changedProperties.has('value') || changedProperties.has('type')) {
+    if (
+      changedProperties.has('value') ||
```

Comment:
Actually this comment made me realize we can use CodeMirror's `setValue` and `setOption` to sync all of these properties without reconstructing the entire CodeMirror instance for every change, so I've done that instead.

BTW to help with this I added a strongly typed version of `ChangedProperties`, which let me do an exhaustive check over all the possibly changed properties. WDYT about upstreaming something like this to lit-element? Note it also has an `Omit<..., keyof LitElement>` so that only the element's own properties are included.

Reply:
We have `PropertyValues<T=any>` right now. The default of `any` is for back-compat, and intended to be used like:

```
update(changedProperties: PropertyValues<this>)
```

https://github.com/Polymer/lit-element/blob/master/src/lib/updating-element.ts#L142

Unfortunately because PropertyValues does the `keyof T` automatically, it doesn't quite admit the `Omit<..., keyof LitElement>` pattern unless we added a second optional type param. Now's the time to change this in 3.0 if we want to.

Reply:
Currently `PropertyValues<this>` doesn't give you any useful type information on `.keys`, it's just `string|number|Symbol`, so it does seem like there is room for improvement in that type. Not sure I understand why we'd need two parameters, the type I have here seems to work well.

Reply:
`PropertyValues<this>` _should_ give types on `.keys()`. I see it working [here](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcAKUItogagQwBsBXYAZwB4AVOAXjnwDtEAaOAeTrmMYGtGIAd0YA+OgCg4UuL2CIIAMw4BbAJYxqbWfKXsxoGMEYATMukzYkAaTlwA-HACy+MBW2KV6zTLke9bHn4hUQAuOEZgADdgKABucXEAY0J8MjMAQTgAb0kGOzCyGChVRgBzeKkAI3zw4mVKmPiAXwTk1LMAITgDI1M4TJypRJrKiAhCYCZmhKQUcywYmFVyTPoMBZwCEnIKdJF4xIhGQoYw9ctlslXw4EEnFwAKAEp4-AA6UuAYB4ByRJ+Xkkjid8DZEGZ6O9tGRnvFxLNUOdFpcumsLIs8ERSJQOmw9gdgfBKmd0bAUVwIndnGBYeJKh8vr9-oD6Z9vj98ACCcciWCIXB6dDaeIgA)

What I mean by the second parameter is the ability to omit the keys from a base class. I added that in the above playground snippet too.

What `PropertyValues<this>` doesn't give you is types for the values on `get`, `set`, `values`, and `entries`, because we don't map the value type. That'd be great to add.